### PR TITLE
[MERGE] (test_)mail/mass_mailing(_sms): improve mailing trace status management

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
PURPOSE

Purpose of this merge is to provide some cleaning in mail application notably
targeting mass mailing and failure management.

SPECIFICATIONS: FAILURE TYPES

Currently there are differences between mail and sms error management
especially when sending them in batch (mass mode). Moreover cancel
(ignored) and error (exception) states meaning is not clear. Finally
some failure types management between mail and sms can be cleaned.

Meaning of ignored / error we want to enforce now is

  * error: there was something wrong at sending and user has an action to
    perform, i.e. server failed -> check its logs;
  * canceled: invalid recipients due to contact information or mailing
    configuration (blacklist, opt out, void or invalid email or phone number).
    In indicates issues linked to records themselves;

In this task we also add failure information granularity on mailing traces
linked to email like what is done currently on SMS. This can be related to
mailing (blacklist, optout, duplicates) or related to recipient (no recipient,
incorrectly formatted).

We also correctly distinguish optout from blacklist when sending SMS.

Detect and store failure types directly on mail.mail: blacklist, optout,
duplicates, missing email, wrong email, ...

State computation is moved from mass mailing to mail. It is also improved as
currently everything is based on "first recipient found". This is globally
valid for mass mailing who uses default recipients (customer). However when
dealing with generic mailing this is not true anymore.

We choose to store and update status on mail.mail records only when having
a single recipient. When several recipients are defined on a given mail.mail
we cannot really decide a status prior to sending it and skip this computation.

Mass mailing override now uses information from mail.mail to update its linked
traces as computation is now delegated to base composer model.

SPECIFICATIONS: TRACE STATUS MANAGEMENT

In this merge we clean the way trace state is managed. Currently it is a
computed field based on several datetime fields. However this generates a
lot of noise in the table as well as unnecessary computation

  * there are several columns (one for each state) storing datetime at
    which status was reached. Generally only 2 or 3 contain relevant
    information;
  * state could be set in code directly to avoid a computed field based on
    many triggers;
  * recomputing it each time a date changes is not necessarily necessary;
  * state value can always be updated manually as this is main done through
    some automated server update (mailgateway, link clicks, ...);

As trace states and its triggers should not be updated manually it is better
to synchronize it in code flow. When there is an exception or update done
through sending or gateway status is updated as well accordingly. Various
datetime fields are also updated at the same time. In order to align with
notification model mail and sms trace status are updated to a classic field.
Only last status update is now kept as there is no need to store the entire
history of status change.

We keep only a datetime for relevant metrics: open, reply and click. Other
datetime bring no real value. Knowing when a trace was in error or bounced
is not necessary. Indeed exception generally indicates a server issue (at
sending), cancel indicates a data issue (at sending) and bounce depends on
customer email server.

Status update is removed as using write_date is sufficient. Once created
traces are updated only when an external event occurs (opened, replied, ...).
It allows to simplify trace model.

We also rename ignored field into canceled to match naming use through mail
and sms.

VARIOUS

Improve code readability by renaming some failure type keys, notably those
used in mail application.

Improve user experience by correctly contextualizing trace display when coming
from a mailing.

Improve user experience of report model build on mailing traces.

See sub commits for more details.

LINKS

Task ID-2377974

### CODE REVIEW

File: addons/mail/models/mail_notification.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 class MailNotification(models.Model):
     _name = 'mail.notification'
-    _table = 'mail_message_res_partner_needaction_rel'
+    _table = 'mail_notification'
```
Comment (seb-odoo) [inline addons/mail/models/mail_notification.py:12]: Finally!
So do you have to still specify the table name explicitly here now? Since it is now the default table name matching model name.

Comment (tde-banana-odoo) [reply addons/mail/models/mail_notification.py:12]: Better explicit than implicit for this one :D

File: addons/mass_mailing/models/mailing_trace.py [hunk 2]
```diff
@@ -146,33 +122,48 @@ def action_view_contact(self):
...
+        traces = self if self else self._find_traces(mail_mail_ids, message_ids)
+        traces.write({'trace_status': 'error', 'trace_status_update': fields.Datetime.now(), 'failure_type': failure_type})
         return traces
 
-    def set_bounced(self, mail_mail_ids=None, mail_message_ids=None):
-        traces = self._get_records(mail_mail_ids, mail_message_ids, [('bounced', '=', False), ('opened', '=', False)])
-        traces.write({'bounced': fields.Datetime.now()})
+    def set_canceled(self, mail_mail_ids=None, message_ids=None):
+        traces = self if self else self._find_traces(mail_mail_ids, message_ids)
+        traces.write({'trace_status': 'cance', 'trace_status_update': fields.Datetime.now()})
```
Comment (KangOl) [inline addons/mass_mailing/models/mailing_trace.py:168]:
```suggestion
        traces.write({'trace_status': 'cancel', 'trace_status_update': fields.Datetime.now()})
```

File: addons/mail/models/mail_message.py [hunk 3]
```diff
@@ -28,37 +28,37 @@ class Message(models.Model):
...
-            # mail
-            "SMTP", "RECIPIENT", "BOUNCE", "UNKNOWN"
-            # sms (SMS addon)
-            'sms_number_missing', 'sms_number_format', 'sms_credit',
-            'sms_server', 'sms_acc'
-            # snailmail (snailmail addon)
-            'sn_credit', 'sn_trial', 'sn_price', 'sn_fields',
-            'sn_format', 'sn_error'
+        # mail
+        "SMTP", "RECIPIENT", "BOUNCE", "UNKNOWN", "mail_missing"
```
Comment (awa-odoo) [inline addons/mail/models/mail_message.py]: "mail_missing" could be confusing IMO, it's not a mail that is missing but an address.
Why not go for something explicit like "recipient_address_missing" or "mail_address_missing" or something?

File: addons/mass_mailing/tests/common.py [hunk 4]
```diff
@@ -104,14 +104,15 @@ def assertMailTraces(self, recipients_info, mailing, records,
                 'MailTrace: email %s (recipient %s, state: %s, record: %s): found %s records (1 expected)' % (email, partner, state, record, len(recipient_trace))
             )
             self.assertTrue(bool(recipient_trace.mail_mail_id_int))
+            if 'failure_type' in recipient_info or state in ('ignored', 'exception', 'canceled', 'bounced'):
```
Comment (awa-odoo) [inline addons/mass_mailing/tests/common.py]: Are you sure it's a "or" here?

File: addons/mass_mailing/models/mailing.py [hunk 5]
```diff
@@ -194,15 +194,15 @@ def _compute_statistics(self):
...
-                COUNT(CASE WHEN s.opened is not null THEN 1 ELSE null END) AS opened,
-                COUNT(CASE WHEN s.clicked is not null THEN 1 ELSE null END) AS clicked,
-                COUNT(CASE WHEN s.replied is not null THEN 1 ELSE null END) AS replied,
-                COUNT(CASE WHEN s.bounced is not null THEN 1 ELSE null END) AS bounced,
-                COUNT(CASE WHEN s.exception is not null THEN 1 ELSE null END) AS failed
+                COUNT(CASE WHEN s.sent_datetime is not null THEN 1 ELSE null END) AS sent,
+                COUNT(CASE WHEN s.trace_status = 'outgoing' THEN 1 ELSE null END) AS scheduled,
+                COUNT(CASE WHEN s.trace_status = 'cancel' THEN 1 ELSE null END) AS canceled,
+                COUNT(CASE WHEN s.trace_status in ('sent', 'open', 'reply') THEN 1 ELSE null END) AS delivered,
+                COUNT(CASE WHEN s.trace_status IN ('open', 'reply') THEN 1 ELSE null END) AS opened,
```
Comment (awa-odoo) [inline addons/mass_mailing/models/mailing.py:201]: "in" or "IN", but not both :D

File: addons/mass_mailing/models/mailing.py [hunk 6]
```diff
@@ -426,25 +426,31 @@ def action_view_clicked(self):
...
+        if view_filter in ('reply', 'bounce'):
+            found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.trace_status == view_filter)
+        elif view_filter == 'open':
+            found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.trace_status in ('open', 'reply'))
+        elif view_filter == 'click':
+            found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.links_click_done)
+        elif view_filter == 'delivered':
+            found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.trace_status in ('sent', 'open', 'reply'))
+        elif view_filter == 'sent':
+            found_traces = self.mailing_trace_ids.filtered(lambda trace: trace.sent_datetime)
```
Comment (awa-odoo) [inline addons/mass_mailing/models/mailing.py:450]: There is a bit of mismatch between all of those, they should be either 
- all verbs (reply, open, bounce, deliver, send)
- all past tense (replied, opened, bounced, delivered, sent)

File: addons/mass_mailing/models/mailing_trace.py [hunk 7]
```diff
@@ -106,34 +97,15 @@ class MailingTrace(models.Model):
         ("mail_optout", "Opted Out"),
         ("mail_dup", "Duplicated Email"),
     ], string='Failure type')
-    state_update = fields.Datetime(compute="_compute_state", string='State Update',
-                                   help='Last state update of the mail',
-                                   store=True)
+    # Link tracking
+    links_click_ids = fields.One2many('link.tracker.click', 'mailing_trace_id', string='Links click')
+    links_click_done = fields.Boolean('Links Clicked', help='If customer clicked on at least one tracked link')
```
Comment (awa-odoo) [inline addons/mass_mailing/models/mailing_trace.py:105]: Feels strange that those 2 are not related in any way, seems like one should be computed based on the other?

File: addons/mass_mailing/models/mailing_trace.py [hunk 8]
```diff
@@ -151,33 +123,37 @@ def action_view_contact(self):
...
-            base_domain = [('id', 'in', self.ids)]
-        if domain:
-            base_domain = ['&'] + domain + base_domain
-        return self.search(base_domain)
-
-    def set_opened(self, mail_mail_ids=None, mail_message_ids=None):
-        traces = self._get_records(mail_mail_ids, mail_message_ids, [('opened', '=', False)])
-        traces.write({'opened': fields.Datetime.now(), 'bounced': False})
+    def set_sent(self, domain=None):
+        traces = self if self else self.search(domain)
```
Comment (awa-odoo) [inline addons/mass_mailing/models/mailing_trace.py:130]: Why not go for something more inclusive?

```py
traces = self
if domain:
    traces |= self.search(domain)
# ...
```

File: addons/test_mass_mailing/tests/test_blacklist_behavior.py [hunk 9]
```diff
@@ -16,10 +18,12 @@ def test_mailing_bounce_w_auto_bl(self):
         mailing.write({'mailing_domain': [('id', 'in', target.ids)]})
 
         # create bounced history of 4 statistics
+        traces = self.env['mailing.trace']
         for idx in range(4):
             new_mailing = mailing.copy()
-            self._create_bounce_trace(new_mailing, target, dt=datetime.datetime.now() - datetime.timedelta(weeks=idx+2))
-            self.gateway_mail_bounce(new_mailing, target)
+            traces += self._create_bounce_trace(new_mailing, target, dt=datetime.datetime.now() - datetime.timedelta(weeks=idx+2))
+            with freeze_time(datetime.datetime.now() - datetime.timedelta(weeks=idx+2)):
```
Comment (awa-odoo) [inline addons/test_mass_mailing/tests/test_blacklist_behavior.py:25]: Isn't it safer to store "datetime.datetime.now()" (or even the whole computed datetime) into a variable instead of calling it twice?

File: addons/mail/models/mail_mail.py [hunk 10]
```diff
@@ -182,7 +182,7 @@ def _postprocess_sent_message(self, success_pids, failure_reason=False, failure_
                     messages = notifications.mapped('mail_message_id').filtered(lambda m: m.is_thread_message())
                     # TDE TODO: could be great to notify message-based, not notifications-based, to lessen number of notifs
                     messages._notify_message_notification_update()  # notify user that we have a failure
-        if not failure_type or failure_type in ['RECIPIENT', 'mail_missing']:  # if we have another error, we want to keep the mail.
+        if not failure_type or failure_type in ['mail_email', 'mail_missing']:  # if we have another error, we want to keep the mail.
```
Comment (awa-odoo) [inline addons/mail/models/mail_mail.py:185]: Commit message states "m_mail" but here it seems to be "mail_email" (same for others apparently).

File: addons/mail/models/mail_mail.py [hunk 11]
```diff
@@ -345,7 +345,7 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
                     notif_msg = _('Error without exception. Probably due do concurrent access update of notification records. Please see with an administrator.')
                     notifs.sudo().write({
                         'notification_status': 'exception',
-                        'failure_type': 'UNKNOWN',
+                        'failure_type': 'error',
```
Comment (awa-odoo) [inline addons/mail/models/mail_mail.py:348]: I actually liked "unknown" more.
"error" is not really a failure type, and is a bit redundant with 'notification_status': 'exception', don't you think?

**Review by awa-odoo — COMMENTED** [review]
Blorg.

Code looks good overall, nice little refactoring / cleaning.
I left my usual annoying naming remarks along with some suggestions.

Cheers!

File: addons/mail/models/mail_mail.py [hunk 12]
```diff
@@ -379,7 +379,10 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
                         processing_pid = None
                     except AssertionError as error:
                         if str(error) == IrMailServer.NO_VALID_RECIPIENT:
-                            failure_type = "RECIPIENT"
+                            if not email.get('email_to'):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_mail.py:382]: That's actually ... not completely False. To be complete we should also check email_cc . Will make an additional commit for this one.

File: addons/mail/models/mail_message.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_message.py]: It was mainly to keep keys short in db. Its label gives an explanation (grepping it returns few occurences in code so easy to find it).

File: addons/mass_mailing/tests/common.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing/tests/common.py]: Well actually we should always have both to have complete tests. The "or" was there mainly when updating tests to detect incomplete tests (missing failure type). As those are tools used mainly by me, well, will probably keep it like that.

File: addons/mass_mailing/models/mailing_trace.py (hunk 7 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing_trace.py:105]: It is computed by hand (aka updated when going through link click code). Purpose is to avoid having a computed based on o2m, quickly inefficient when dealing with a model with high volume.

Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing_trace.py:105]: I made the same move for status, it is actually one of the main reason of this PR: simpler to update manually as we go through some flows, no need to keep computed fields as people should not update "dependencies" manually.

File: addons/mail/models/mail_mail.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_mail.py:385]: As it requires some further exploration, added a comment and will create a task to investigate.

File: addons/mail/models/mail_mail.py (hunk 11 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_mail.py:351]: Seems legit.

File: addons/mass_mailing/models/mailing.py [hunk 13]
```diff
@@ -194,15 +194,15 @@ def _compute_statistics(self):
...
-                COUNT(CASE WHEN s.sent is not null THEN 1 ELSE null END) AS sent,
-                COUNT(CASE WHEN s.scheduled is not null AND s.sent is null AND s.exception is null AND s.ignored is null AND s.bounced is null THEN 1 ELSE null END) AS scheduled,
-                COUNT(CASE WHEN s.scheduled is not null AND s.sent is null AND s.exception is null AND s.ignored is not null THEN 1 ELSE null END) AS ignored,
-                COUNT(CASE WHEN s.sent is not null AND s.exception is null AND s.bounced is null THEN 1 ELSE null END) AS delivered,
-                COUNT(CASE WHEN s.opened is not null THEN 1 ELSE null END) AS opened,
-                COUNT(CASE WHEN s.clicked is not null THEN 1 ELSE null END) AS clicked,
-                COUNT(CASE WHEN s.replied is not null THEN 1 ELSE null END) AS replied,
-                COUNT(CASE WHEN s.bounced is not null THEN 1 ELSE null END) AS bounced,
-                COUNT(CASE WHEN s.exception is not null THEN 1 ELSE null END) AS failed
+                COUNT(CASE WHEN s.sent_datetime is not null THEN 1 ELSE null END) AS sent,
```
Comment (nseinlet) [inline addons/mass_mailing/models/mailing.py:197]:
```suggestion
                COUNT(s.sent_datetime) AS sent,
```

File: addons/mass_mailing/models/mailing.py [hunk 14]
```diff
@@ -194,15 +194,15 @@ def _compute_statistics(self):
...
+                COUNT(CASE WHEN s.trace_status = 'outgoing' THEN 1 ELSE null END) AS scheduled,
```
Comment (KangOl) [inline addons/mass_mailing/models/mailing.py:198]:
```suggestion
                COUNT(s.trace_status) FILTER (WHERE s.trace_status = 'outgoing') AS scheduled,
```

File: addons/mass_mailing/models/mailing.py [hunk 15]
```diff
@@ -194,15 +194,15 @@ def _compute_statistics(self):
...
-                COUNT(CASE WHEN s.clicked is not null THEN 1 ELSE null END) AS clicked,
-                COUNT(CASE WHEN s.replied is not null THEN 1 ELSE null END) AS replied,
-                COUNT(CASE WHEN s.bounced is not null THEN 1 ELSE null END) AS bounced,
-                COUNT(CASE WHEN s.exception is not null THEN 1 ELSE null END) AS failed
+                COUNT(s.sent_datetime) AS sent,
+                COUNT(s.trace_status) FILTER (WHERE s.trace_status = 'outgoing') AS scheduled,
+                COUNT(s.trace_status) FILTER (WHERE s.trace_status = 'cancel') AS canceled,
+                COUNT(s.trace_status) FILTER (WHERE s.trace_status in ('sent', 'open', 'reply')) AS delivered,
+                COUNT(s.trace_status) FILTER (WHERE s.trace_status in ('open', 'reply')) AS opened,
+                COUNT(s.links_click_done) AS clicked,
```
Comment (KangOl) [inline addons/mass_mailing/models/mailing.py:202]: Will count all not null values, even `false`.
```suggestion
                COUNT(s.links_click_done) FILTER (WHERE s.links_click_done) AS clicked,
```

File: addons/mail/wizard/mail_compose_message.py [hunk 16]
```diff
@@ -283,6 +310,30 @@ def send_mail(self, auto_commit=False):
...
+        for record in self:
+            model = self.env['ir.model']._get(record.model or 'mail.message')
+            model_name = model.name or ''
+            template_name = "%s: %s" % (model_name, tools.ustr(record.subject))
+            values = {
+                'name': template_name,
+                'subject': record.subject or False,
+                'body_html': record.body or False,
+                'model_id': model.id or False,
+                'attachment_ids': [Command.set([att.id for att in record.attachment_ids])],
```
Comment (Feyensv) [inline addons/mail/wizard/mail_compose_message.py:325]:
```suggestion
                'attachment_ids': [Command.set(record.attachment_ids.ids)],
```

File: addons/mail/wizard/mail_compose_message.py [hunk 17]
```diff
@@ -161,6 +158,15 @@ def _compute_render_model(self):
         for composer in self:
             composer.render_model = composer.model
 
+    # Onchanges
+
+    @api.onchange('template_id')
+    def onchange_template_id_wrapper(self):
```
Comment (Feyensv) [inline addons/mail/wizard/mail_compose_message.py:164]:
```suggestion
    def _onchange_template_id_wrapper(self):
```
While you're moving it, apply guidelines fully :D

**Review by Feyensv — APPROVED** [review]
Two dummy comments noticed on the fly :duck:
Interesting changes, safe diff for sale :ok_hand:

File: addons/mail/wizard/mail_compose_message.py [hunk 18]
```diff
@@ -412,15 +402,115 @@ def get_mail_values(self, res_ids):
...
+            elif done_emails is not None:
+                done_emails.append(mail_to)
+
+        return mail_values_dict
+
+    def _get_blacklist_record_ids(self, mail_values_dict):
+        blacklisted_rec_ids = set()
+        if self.composition_mode == 'mass_mail' and issubclass(type(self.env[self.model]), self.pool['mail.thread.blacklist']):
+            self.env['mail.blacklist'].flush(['email'])
+            self._cr.execute("SELECT email FROM mail_blacklist WHERE active=true")
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_compose_message.py:497]: This is... pretty brutal :D (I know it's copy/pasted)

File: addons/mass_mailing/models/mailing.py [hunk 19]
```diff
@@ -442,10 +442,14 @@ def action_view_traces_sent(self):
...
     def _action_view_traces_filtered(self, view_filter):
         action = self.env["ir.actions.actions"]._for_xml_id("mass_mailing.mailing_trace_action")
-        action['name'] = _('%s Traces') % (self.name)
+        action['name'] = _('Sent Mailings')
         action['context'] = {'search_default_mass_mailing_id': self.id,}
         filter_key = 'search_default_filter_%s' % (view_filter)
         action['context'][filter_key] = True
+        action['views'] = [
+            (self.env.ref('mass_mailing.mailing_trace_view_tree_mail').id, 'tree'),
+            (self.env.ref('mass_mailing.mailing_trace_view_form').id, 'form')
```
Comment (awa-odoo) [inline addons/mass_mailing/models/mailing.py:451]: If you want the default form, I don't think you have to specify it (or you could do something like `0, 'form'`)

File: addons/mass_mailing/report/mailing_trace_report.py [hunk 20]
```diff
@@ -19,6 +19,7 @@ class MailingTraceReport(models.Model):
         string='Status', readonly=True)
     email_from = fields.Char('From', readonly=True)
     # traces
+    scheduled = fields.Integer(readonly=True)
```
Comment (awa-odoo) [inline addons/mass_mailing/report/mailing_trace_report.py:22]: Commit message says:
> Improve user experience by improve report model build on mailing traces.

Sad AWA is sad.

**Review by awa-odoo — COMMENTED** [review]
Hello @tde-banana-odoo 

Re-read the whole thing.
Looking good! I added some final annoying polishing comments as always.

I noticed that a lot of your commits re-write the previous ones, so I want to apologize and deny all responsibility if my comments create any kind of rebase hell for you ^^

Cheers!

File: addons/mass_mailing/wizard/mail_compose_message.py [hunk 21]
```diff
@@ -57,30 +57,53 @@ def get_mail_values(self, res_ids):
...
+                    mail_values['state'] = 'cancel'
+                    error_code = 'mail_dup'
+                # void of falsey values -> error
+                elif not mail_to:
+                    mail_values['state'] = 'exception'
+                    error_code = 'mail_email_missing'
+                elif not mail_to_normalized or not email_re.findall(mail_to):
+                    mail_values['state'] = 'exception'
+                    error_code = "RECIPIENT"
                 elif seen_list is not None:
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/wizard/mail_compose_message.py]: Actually this code is moved in a next commit directly into mail, so ... code will be different (and simpler in mass mailing).

File: addons/mass_mailing/models/mailing_trace.py [hunk 22]
```diff
@@ -62,39 +62,30 @@ class MailingTrace(models.Model):
...
+    open_datetime = fields.Datetime('Opened On')
+    reply_datetime = fields.Datetime('Replied On')
+    trace_status = fields.Selection(selection=[
+        ('outgoing', 'Outgoing'),
+        ('sent', 'Sent'),
+        ('open', 'Opened'),
+        ('reply', 'Replied'),
+        ('bounce', 'Bounced'),
+        ('error', 'Exception'),
+        ('cancel', 'Canceled')], string='Status', default='outgoing')
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing_trace.py:90]: Send is ugly :D . Actually I don't like open and reply. But meh, marre de tout changer.

File: addons/mass_mailing/models/mailing_trace.py [hunk 23]
```diff
@@ -151,33 +123,37 @@ def action_view_contact(self):
...
+        traces = self + (self.search(domain) if domain else self.env['mailing.trace'])
+        traces.write({'trace_status': 'bounce'})
         return traces
 
-    def set_replied(self, mail_mail_ids=None, mail_message_ids=None):
-        traces = self._get_records(mail_mail_ids, mail_message_ids, [('replied', '=', False)])
-        traces.write({'replied': fields.Datetime.now()})
+    def set_failed(self, domain=None, failure_type=False):
+        traces = self + (self.search(domain) if domain else self.env['mailing.trace'])
+        traces.write({'trace_status': 'error', 'failure_type': failure_type})
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing_trace.py:156]: Normally always set. And if you set another error without type, old type does not makes sense anymore anyway.

File: addons/mass_mailing/tests/common.py [hunk 24]
```diff
@@ -200,15 +205,16 @@ def _create_bounce_trace(cls, mailing, record, dt=None):
...
-        trace = cls.env['mailing.trace'].create({
-            'mass_mailing_id': mailing.id,
-            'model': record._name,
-            'res_id': record.id,
-            'bounced': dt,
-            # TDE FIXME: improve this with a mail-enabled heuristics
-            'email': trace_email,
-            'message_id': '<%5f@gilbert.boitempomils>' % randomized,
-        })
+        with patch.object(Cursor, 'now', lambda *args, **kwargs: dt):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/tests/common.py:208]: Actually now on Cursor does not use datetime but instead uses postgresql now. Is probably worth a note :D .

File: addons/mass_mailing_sms/tests/common.py [hunk 25]
```diff
@@ -31,7 +31,7 @@ def assertSMSTraces(self, recipients_info, mailing, records,
           # TRACE
           'partner': res.partner record (may be empty),
           'number': number used for notification (may be empty, computed based on partner),
-          'state': outgoing / sent / ignored / bounced / exception / opened (outgoing by default),
+          'trace_status': outgoing / sent / cancel / bounce / error / opened (sent by default),
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing_sms/tests/common.py:34]: It just considers that sent is default value in test. Not really clear but hey, test tools.

File: addons/test_mass_mailing/tests/test_blacklist_behavior.py [hunk 26]
```diff
@@ -42,15 +49,19 @@ def test_mailing_bounce_w_auto_bl_partner_duplicates(self):
...
         target = self.env['mailing.test.blacklist'].browse(self.target_rec.ids)
 
         # create bounced history of 4 statistics
+        traces = self.env['mailing.trace']
         for idx in range(4):
             new_mailing = mailing.copy()
-            self._create_bounce_trace(new_mailing, target, dt=datetime.datetime.now() - datetime.timedelta(weeks=idx+2))
-            self.gateway_mail_bounce(new_mailing, target, bounce_base_values)
+            new_dt = datetime.datetime.now() - datetime.timedelta(weeks=idx+2)
+            with freeze_time(new_dt), patch.object(Cursor, 'now', lambda *args, **kwargs: new_dt):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mass_mailing/tests/test_blacklist_behavior.py:62]: Added a comment like the other cursor patch

File: addons/mass_mailing/models/mailing.py (hunk 19 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing.py:451]: Makes inherit in _sms more explicit in my opinion ... otherwise indeed it is default value. But here I want it to be explicit.

File: addons/mail/wizard/mail_compose_message.py [hunk 27]
```diff
@@ -412,15 +402,115 @@ def get_mail_values(self, res_ids):
...
+                'mail_to': mail_to,
+                'mail_to_normalized': [
+                    tools.email_normalize(mail)
+                    for mail in mail_to
+                    if tools.email_normalize(mail)
+                ]
+            }
+        return recipients_info
+
+    def _process_state(self, mail_values_dict):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:451]: A bit marre of this code. Not optimal code, but a cleaning of this belongs to a new task (and it is planned to do it). Composer is quite old fashioned. So as discussed on discord: i keep it as it is currently, to be cleaned in upcoming version.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
